File: android/src/main/java/com/wxwx/flutter_alibc/FlutterAlibcHandle.java
package com.wxwx.flutter_alibc;

import android.util.Log;
import android.webkit.WebChromeClient;
import android.webkit.WebViewClient;
import android.widget.Toast;
import android.content.Intent;
import com.ali.auth.third.core.model.Session;
import com.alibaba.baichuan.trade.biz.context.AlibcResultType;
import com.alibaba.baichuan.android.trade.AlibcTrade;
import com.alibaba.baichuan.android.trade.callback.AlibcTradeCallback;
import com.alibaba.baichuan.android.trade.model.AlibcShowParams;
import com.alibaba.baichuan.android.trade.model.OpenType;
import com.alibaba.baichuan.android.trade.page.AlibcBasePage;
import com.alibaba.baichuan.android.trade.page.AlibcDetailPage;
import com.alibaba.baichuan.android.trade.page.AlibcMyCartsPage;
import com.alibaba.baichuan.android.trade.page.AlibcShopPage;
import com.alibaba.baichuan.trade.biz.applink.adapter.AlibcFailModeType;
import com.alibaba.baichuan.trade.biz.context.AlibcTradeResult;
import com.alibaba.baichuan.trade.biz.core.taoke.AlibcTaokeParams;
import com.alibaba.baichuan.trade.biz.login.AlibcLogin;
import com.alibaba.baichuan.trade.biz.login.AlibcLoginCallback;
import com.alibaba.baichuan.android.trade.AlibcTradeSDK;
import com.alibaba.baichuan.android.trade.callback.AlibcTradeInitCallback;
import com.alibaba.baichuan.trade.common.utils.AlibcLogger;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.MethodCall;
import java.util.HashMap;
import android.app.AlertDialog;
import static com.wxwx.flutter_alibc.PluginConstants.*;
import static com.wxwx.flutter_alibc.PluginUtil.*;
import com.wxwx.flutter_alibc.web.WebViewActivity;
import java.util.Map;

/**
 * @Author karedem
 * @Date 2019/9/7 19:55
 * @Description 接口处理者
**/
public class FlutterAlibcHandle{

    private static FlutterAlibcHandle handle;
    private Registrar register;

    //第一次调用getInstance register不能为空
    public static FlutterAlibcHandle getInstance(Registrar register){
        if (handle == null){
            synchronized (FlutterAlibcHandle.class){
                handle = new FlutterAlibcHandle();
                handle.register = register;
            }
        }
        return handle;
    }

    /**
     * 初始化阿里百川
     * @param call
     * @param result
     */
    public void initAlibc(MethodCall call, Result result){
        AlibcTradeSDK.asyncInit(register.activity().getApplication(), new AlibcTradeInitCallback() {
            @Override
            public void onSuccess() {
                result.success(PluginResponse.success(null).toMap());
            }
            @Override
            public void onFailure(int code, String msg) {
                result.success(new PluginResponse(Integer.toString(code), msg, null).toMap());
            }
        });
    }

    /**
     * 登陆淘宝
     * @param result
     */
    public void loginTaoBao(Result result){
        final AlibcLogin alibcLogin = AlibcLogin.getInstance();
        if (alibcLogin.isLogin()){
            Session session = AlibcLogin.getInstance().getSession();
            Map<String, Object> userInfo = new HashMap<>();
            userInfo.put("nick", session.nick);
            userInfo.put("avatarUrl", session.avatarUrl);
            userInfo.put("openId", session.openId);
            userInfo.put("openSid", session.openSid);
            userInfo.put("topAccessToken", session.topAccessToken);
            userInfo.put("topAuthCode", session.topAuthCode);
            result.success(PluginResponse.success(userInfo).toMap());
            return;
        }
        alibcLogin.showLogin(new AlibcLoginCallback() {
            @Override
            public void onSuccess(int loginResult, String openId, String userNick) {
                Map<String, Object> userInfo = new HashMap<>();
                Session session = AlibcLogin.getInstance().getSession();
                userInfo.put("nick", session.nick);
                userInfo.put("avatarUrl", session.avatarUrl);
                userInfo.put("openId", session.openId);
                userInfo.put("openSid", session.openSid);
                userInfo.put("topAccessToken", session.topAccessToken);
                userInfo.put("topAuthCode", session.topAuthCode);
                result.success(PluginResponse.success(userInfo).toMap());
            }
            @Override
            public void onFailure(int code, String msg) {
                // code：错误码  msg： 错误信息
                result.success(new PluginResponse(Integer.toString(code), msg, null).toMap());
            }
        });
    }

    /**
     * 登出
     * @param result
     */
    public void loginOut(Result result){
        AlibcLogin alibcLogin = AlibcLogin.getInstance();
        alibcLogin.logout(new AlibcLoginCallback() {
            @Override
            public void onSuccess(int loginResult, String openId, String userNick) {
            }
            @Override
            public void onFailure(int code, String msg) {
                // code：错误码  msg： 错误信息
            }
        });
    }

    /**
     * 淘宝授权登陆  获取access_token
     *  官方说明文档 {https://open.taobao.com/doc.htm?docId=118&docType=1}
     * @param call
     * @param result
     */
    public void taoKeLogin(MethodCall call, Result result){
        HashMap<String, Object> map = (HashMap<String, Object>)call.arguments;
        String url = call.argument("url");
        WebViewActivity.setCallBack(new WebViewActivity.CallBack() {
            @Override
            public void success(String accessToken) {
                Map<String,String> resMap = new HashMap();
                resMap.put("accessToken", accessToken);
                result.success(resMap);
            }

            @Override
            public void failed(String errorMsg) {
                Map<String,String> resMap = new HashMap();
                resMap.put("accessToken", "");
                result.success(resMap);
            }
        });
        Intent intent = new Intent(register.activity(), WebViewActivity.class);
        intent.putExtra("url", url);
        intent.putExtra("arguments", map);
        register.activity().startActivity(intent);
    }

    /**
     * 通过URL方式打开淘宝
     * @param call
     * @param result
     */
    public void openByUrl(MethodCall call, Result result){
        AlibcShowParams showParams = new AlibcShowParams();
        AlibcTaokeParams taokeParams = new AlibcTaokeParams("","","");

        showParams.setBackUrl(call.argument(key_BackUrl));

        if (call.argument(key_OpenType) != null){
            System.out.println("openType" + call.argument(key_OpenType));
            showParams.setOpenType(getOpenType(""+call.argument(key_OpenType)));
        }
        if (call.argument(key_ClientType) != null){
            System.out.println("clientType " + call.argument(key_ClientType));
            showParams.setClientType(getClientType(""+call.argument(key_ClientType)));
        }
        if (call.argument("taokeParams") != null){
            taokeParams  = getTaokeParams(call.argument("taokeParams"));
        }
        if ("false".equals(call.argument("isNeedCustomNativeFailMode"))){
            showParams.setNativeOpenFailedMode(AlibcFailModeType.AlibcNativeFailModeNONE);
        }else if (call.argument(key_NativeFailMode) != null){
            showParams.setNativeOpenFailedMode(getFailModeType(""+call.argument(key_NativeFailMode)));
        }

        Map<String, String> trackParams = new HashMap<>();
        String url = call.argument("url");
        // 以显示传入url的方式打开页面（第二个参数是套件名称）
        AlibcTrade.openByUrl(register.activity(), "", url, null,
                new WebViewClient(), new WebChromeClient(), showParams,
                taokeParams, trackParams, new AlibcTradeCallback() {
                    @Override
                    public void onTradeSuccess(AlibcTradeResult tradeResult) {
                        Map<String, Object> results = new HashMap<>();
                        if (AlibcResultType.TYPECART == tradeResult.resultType){
                            results.put("type", 1);
                        }else if (AlibcResultType.TYPEPAY == tradeResult.resultType){
                            results.put("type", 0);
                            results.put("payFailedOrders", tradeResult.payResult.payFailedOrders);
                            results.put("paySuccessOrders", tradeResult.payResult.paySuccessOrders);
                        }
                        result.success(PluginResponse.success(results).toMap());
                    }
                    @Override
                    public void onFailure(int code, String msg) {
                        result.success(new PluginResponse(Integer.toString(code), msg, null).toMap());
                    }
                });
    }

    /**
     * 打开商店
     * @param call
     * @param result
     */
    public void openShop(MethodCall call, Result result){
        AlibcBasePage page = new AlibcShopPage(call.argument("shopId"));
        openByBizCode(page, "shop", call, result);
    }

    /**
     * 打开购物车
     * @param result
     */
    public void openCart(MethodCall call, Result result){
        AlibcBasePage page = new AlibcMyCartsPage();
        openByBizCode(page, "cart",call, result);
    }

    /**
     * 打开商品详情
     * @param call   call.argument["itemID"]  详情id
     * @param result
     */
    public void openItemDetail(MethodCall call, Result result){
        AlibcBasePage page = new AlibcDetailPage(call.argument("itemID"));
        openByBizCode(page, "detail", call, result);
    }

    private void openByBizCode(AlibcBasePage page,String type, MethodCall call, Result result){
        AlibcShowParams showParams = new AlibcShowParams();
        AlibcTaokeParams taokeParams = new AlibcTaokeParams("", "", "");

        showParams.setBackUrl(call.argument(key_BackUrl));

        if (call.argument(key_OpenType) != null){
            showParams.setOpenType(getOpenType(""+call.argument(key_OpenType)));
        }
        if (call.argument(key_ClientType) != null){
            showParams.setClientType(getClientType(""+call.argument(key_ClientType)));
        }
        if (call.argument("taokeParams") != null){
            taokeParams  = getTaokeParams(call.argument("taokeParams"));
        }

        if ("false".equals(call.argument("isNeedCustomNativeFailMode"))){
            showParams.setNativeOpenFailedMode(AlibcFailModeType.AlibcNativeFailModeNONE);
        }else if (call.argument(key_NativeFailMode) != null){
            showParams.setNativeOpenFailedMode(getFailModeType(""+call.argument(key_NativeFailMode)));
        }

        Map<String, String> trackParams = new HashMap<>();
        AlibcTrade.openByBizCode(register.activity(), page, null, new WebViewClient(),
                new WebChromeClient(), type, showParams, taokeParams,
                trackParams, new AlibcTradeCallback() {
                    @Override
                    public void onTradeSuccess(AlibcTradeResult tradeResult) {
                        Map<String, Object> results = new HashMap<>();
                        if (AlibcResultType.TYPECART == tradeResult.resultType){
                            results.put("type", 1);
                        }else if (AlibcResultType.TYPEPAY == tradeResult.resultType){
                            results.put("type", 0);
                            results.put("payFailedOrders", tradeResult.payResult.payFailedOrders);
                            results.put("paySuccessOrders", tradeResult.payResult.paySuccessOrders);
                        }
                        result.success(PluginResponse.success(results).toMap());
                    }
                    @Override
                    public void onFailure(int code, String msg) {
                        // 失败回调信息
                        result.success(new PluginResponse(Integer.toString(code), msg, null).toMap());
                    }
                });
    }

    /**
     * 设置淘客打点策略 是否异步
     * @param call
     */
    public void syncForTaoke(MethodCall call){
        AlibcTradeSDK.setSyncForTaoke(call.argument("isSync"));
    }

    /**
     * TODO
     * @param call
     */
    public void useAlipayNative(MethodCall call){
        AlibcTradeSDK.setShouldUseAlipay(call.argument("isNeed"));
    }


}

File: android/src/main/kotlin/com/wxwx/flutter_alibc/FlutterAlibcHandle.kt
package com.wxwx.flutter_alibc
import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.Intent
import android.webkit.WebChromeClient
import android.webkit.WebViewClient
import com.alibaba.baichuan.android.trade.AlibcTrade
import com.alibaba.baichuan.android.trade.AlibcTradeSDK
import com.alibaba.baichuan.android.trade.callback.AlibcTradeCallback
import com.alibaba.baichuan.android.trade.callback.AlibcTradeInitCallback
import com.alibaba.baichuan.android.trade.model.AlibcShowParams
import com.alibaba.baichuan.android.trade.page.AlibcBasePage
import com.alibaba.baichuan.android.trade.page.AlibcDetailPage
import com.alibaba.baichuan.android.trade.page.AlibcMyCartsPage
import com.alibaba.baichuan.android.trade.page.AlibcShopPage
import com.alibaba.baichuan.trade.biz.applink.adapter.AlibcFailModeType
import com.alibaba.baichuan.trade.biz.context.AlibcResultType
import com.alibaba.baichuan.trade.biz.context.AlibcTradeResult
import com.alibaba.baichuan.trade.biz.core.taoke.AlibcTaokeParams
import com.alibaba.baichuan.trade.biz.login.AlibcLogin
import com.alibaba.baichuan.trade.biz.login.AlibcLoginCallback
import com.wxwx.flutter_alibc.web.WebViewActivity
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.*

class FlutterAlibcHandle(var methodChannel: MethodChannel?){
    var activity:Activity? = null
    fun disposed(){
        this.methodChannel = null
        this.activity = null
    }

    /**
     * 初始化阿里百川
     * @param call
     * @param result
     */
    fun initAlibc(result: MethodChannel.Result){
        AlibcTradeSDK.asyncInit(activity!!.application, object : AlibcTradeInitCallback {
            override fun onSuccess() {
                result.success(PluginResponse.success(null).toMap())
            }

            override fun onFailure(code: Int, msg: String?) {
                result.success(PluginResponse(code.toString(), msg, null).toMap())
            }
        })
    }

    /**
     * 登陆淘宝
     * @param result
     */
    fun loginTaoBao(){
        val alibcLogin = AlibcLogin.getInstance()
        if(alibcLogin.isLogin){
            val session = alibcLogin.session
            val userInfo: MutableMap<String, Any> = HashMap()
            userInfo["nick"] = session.nick
            userInfo["avatarUrl"] = session.avatarUrl
            userInfo["openId"] = session.openId
            userInfo["openSid"] = session.openSid
            userInfo["topAccessToken"] = session.topAccessToken
            userInfo["topAuthCode"] = session.topAuthCode
            methodChannel!!.invokeMethod("AlibcTaobaoLogin", PluginResponse.success(userInfo).toMap())
            return
        }
        alibcLogin.showLogin(object : AlibcLoginCallback {
            override fun onSuccess(loginResult: Int, openId: String?, nickName: String?) {
                val userInfo: MutableMap<String, Any> = HashMap()
                val session = alibcLogin.session
                userInfo["nick"] = session.nick
                userInfo["avatarUrl"] = session.avatarUrl
                userInfo["openId"] = session.openId
                userInfo["openSid"] = session.openSid
                userInfo["topAccessToken"] = session.topAccessToken
                userInfo["topAuthCode"] = session.topAuthCode
                methodChannel!!.invokeMethod("AlibcTaobaoLogin", PluginResponse.success(userInfo).toMap())

            }

            override fun onFailure(code: Int, msg: String?) {
                // code：错误码  msg： 错误信息
                methodChannel!!.invokeMethod("AlibcTaobaoLogin", PluginResponse(code.toString(), msg, null).toMap())
            }
        })
    }

    /**
     * 登出
     * @param result
     */
    fun logoutTaoBao(result: MethodChannel.Result){
        val alibcLogin = AlibcLogin.getInstance()
        alibcLogin.logout(object : AlibcLoginCallback {
            override fun onSuccess(p0: Int, p1: String?, p2: String?) {
                result.success(PluginResponse.success(null).toMap())
            }

            override fun onFailure(code: Int, msg: String?) {
                result.success(PluginResponse(code.toString(), msg, null).toMap())
            }
        })
    }

    /**
     * 淘宝授权登陆  获取access_token
     *  官方说明文档 {https://open.taobao.com/doc.htm?docId=118&docType=1}
     * @param call
     * @param result
     */
    fun taoKeLogin(call: MethodCall){
        val map = call.arguments as HashMap<String, Any>
        val url = call.argument<String>("url")
        WebViewActivity.callBack = object : WebViewActivity.Callback {
            override fun success(accessToken: String?) {
                val resMap: HashMap<String, Any?> = HashMap<String, Any?>()
                resMap["accessToken"] = accessToken
                methodChannel!!.invokeMethod("AlibcTaokeLogin", resMap)
            }

            override fun failed(errorMsg: String?) {
                val resMap: HashMap<String, Any?> = HashMap<String, Any?>()
                resMap["accessToken"] = ""
                methodChannel!!.invokeMethod("AlibcTaokeLogin", resMap)
            }
        }
        val intent = Intent(activity!!, WebViewActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        intent.putExtra("url", url)
        intent.putExtra("arguments", map)
        activity!!.startActivity(intent)
    }

    /**
     * 淘宝授权登陆  获取code 授权码
     *  官方说明文档 {https://open.taobao.com/doc.htm?docId=118&docType=1}
     * @param call
     * @param result
     */
    fun taoKeLoginForCode(call: MethodCall){
        val map = call.arguments as HashMap<*, *>
        val url = call.argument<String>("url")
        WebViewActivity.callBack = object : WebViewActivity.Callback {
            override fun success(accessToken: String?) {
                val resMap: HashMap<String, Any?> = HashMap<String, Any?>()
                resMap["code"] = accessToken
                methodChannel!!.invokeMethod("AlibcTaokeLoginForCode", resMap)
            }

            override fun failed(errorMsg: String?) {
                val resMap: HashMap<String, Any?> = HashMap<String, Any?>()
                resMap["code"] = ""
                methodChannel!!.invokeMethod("AlibcTaokeLoginForCode", resMap)
            }
        }
        val intent = Intent(activity!!, WebViewActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        intent.putExtra("url", url)
        intent.putExtra("arguments", map)
        activity!!.startActivity(intent)
    }

    /**
     * 通过URL方式打开淘宝
     * @param call
     * @param result
     */
    fun openByUrl(call: MethodCall, result: MethodChannel.Result){
        val showParams = AlibcShowParams()
        var taokeParams: AlibcTaokeParams? = AlibcTaokeParams("", "", "")

        showParams.backUrl = call.argument(PluginConstants.key_BackUrl)

        if (call.argument<Any?>(PluginConstants.key_OpenType) != null) {
            println("openType" + call.argument<Any>(PluginConstants.key_OpenType))
            showParams.openType = PluginUtil.getOpenType("" + call.argument<Any>(PluginConstants.key_OpenType))
        }
        if (call.argument<Any?>(PluginConstants.key_ClientType) != null) {
            println("clientType " + call.argument<Any>(PluginConstants.key_ClientType))
            showParams.clientType = PluginUtil.getClientType("" + call.argument<Any>(PluginConstants.key_ClientType))
        }
        if (call.argument<Any?>("taokeParams") != null) {
            taokeParams = PluginUtil.getTaokeParams(call.argument<Map<String?, Any?>>("taokeParams")!!)
        }
        if ("false" == call.argument<Any>("isNeedCustomNativeFailMode")) {
            showParams.nativeOpenFailedMode = AlibcFailModeType.AlibcNativeFailModeNONE
        } else if (call.argument<Any?>(PluginConstants.key_NativeFailMode) != null) {
            showParams.nativeOpenFailedMode = PluginUtil.getFailModeType("" + call.argument<Any>(PluginConstants.key_NativeFailMode))
        }

        val trackParams: Map<String, String> = HashMap()
        val url = call.argument<String>("url")
        // 以显示传入url的方式打开页面（第二个参数是套件名称）
        // 以显示传入url的方式打开页面（第二个参数是套件名称）
        AlibcTrade.openByUrl(activity, "", url, null,
                WebViewClient(), WebChromeClient(), showParams,
                taokeParams, trackParams, object : AlibcTradeCallback {
            override fun onTradeSuccess(tradeResult: AlibcTradeResult) {
                val results: MutableMap<String, Any> = HashMap()
                if (AlibcResultType.TYPECART == tradeResult.resultType) {
                    results["type"] = 1
                } else if (AlibcResultType.TYPEPAY == tradeResult.resultType) {
                    results["type"] = 0
                    results["payFailedOrders"] = tradeResult.payResult.payFailedOrders
                    results["paySuccessOrders"] = tradeResult.payResult.paySuccessOrders
                }
                result.success(PluginResponse.success(results).toMap())
            }

            override fun onFailure(code: Int, msg: String) {
                result.success(PluginResponse(code.toString(), msg, null).toMap())
            }
        })
    }

    /**
     * 打开商店
     * @param call
     * @param result
     */
    fun openShop(call: MethodCall, result: MethodChannel.Result) {
        val page: AlibcBasePage = AlibcShopPage(call.argument("shopId"))
        openByBizCode(page, "shop", call, result)
    }

    /**
     * 打开购物车
     * @param result
     */
    fun openCart(call: MethodCall, result: MethodChannel.Result) {
        val page: AlibcBasePage = AlibcMyCartsPage()
        openByBizCode(page, "cart", call, result)
    }

    /**
     * 打开商品详情
     * @param call   call.argument["itemID"]  详情id
     * @param result
     */
    fun openItemDetail(call: MethodCall, result: MethodChannel.Result) {
        val page: AlibcBasePage = AlibcDetailPage(call.argument("itemID"))
        openByBizCode(page, "detail", call, result)
    }

    private fun openByBizCode(page: AlibcBasePage, type: String, call: MethodCall, result: MethodChannel.Result) {
        val showParams = AlibcShowParams()
        var taokeParams: AlibcTaokeParams? = AlibcTaokeParams("", "", "")
        showParams.backUrl = call.argument(PluginConstants.key_BackUrl)
        if (call.argument<Any?>(PluginConstants.key_OpenType) != null) {
            showParams.openType = PluginUtil.getOpenType("" + call.argument<Any>(PluginConstants.key_OpenType))
        }
        if (call.argument<Any?>(PluginConstants.key_ClientType) != null) {
            showParams.clientType = PluginUtil.getClientType("" + call.argument<Any>(PluginConstants.key_ClientType))
        }
        if (call.argument<Any?>("taokeParams") != null) {
            taokeParams = PluginUtil.getTaokeParams(call.argument<Map<String?, Any?>>("taokeParams")!!)
        }
        if ("false" == call.argument<Any>("isNeedCustomNativeFailMode")) {
            showParams.nativeOpenFailedMode = AlibcFailModeType.AlibcNativeFailModeNONE
        } else if (call.argument<Any?>(PluginConstants.key_NativeFailMode) != null) {
            showParams.nativeOpenFailedMode = PluginUtil.getFailModeType("" + call.argument<Any>(PluginConstants.key_NativeFailMode))
        }
        val trackParams: Map<String, String> = HashMap()
        AlibcTrade.openByBizCode(activity, page, null, WebViewClient(),
                WebChromeClient(), type, showParams, taokeParams,
                trackParams, object : AlibcTradeCallback {
            override fun onTradeSuccess(tradeResult: AlibcTradeResult) {
                val results: MutableMap<String, Any> = HashMap()
                if (AlibcResultType.TYPECART == tradeResult.resultType) {
                    results["type"] = 1
                } else if (AlibcResultType.TYPEPAY == tradeResult.resultType) {
                    results["type"] = 0
                    results["payFailedOrders"] = tradeResult.payResult.payFailedOrders
                    results["paySuccessOrders"] = tradeResult.payResult.paySuccessOrders
                }
                result.success(PluginResponse.success(results).toMap())
            }

            override fun onFailure(code: Int, msg: String) {
                // 失败回调信息
                result.success(PluginResponse(code.toString(), msg, null).toMap())
            }
        })
    }

    /**
     * 设置淘客打点策略 是否异步
     * @param call
     */
    fun syncForTaoke(call: MethodCall) {
        AlibcTradeSDK.setSyncForTaoke(call.argument<Boolean>("isSync")!!)
    }

    fun useAlipayNative(call: MethodCall) {
        AlibcTradeSDK.setShouldUseAlipay(call.argument<Boolean>("isNeed")!!)
    }
}